### ISSUE TITLE
"Invalid prop `data-headlessui-state` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props."

### ISSUE THREAD

**Issue by tordans** [tailwindlabs/headlessui#3351]
**What package within Headless UI are you using?**

`"@headlessui/react": "0.0.0-insiders.fbad6a9",`

**What browser are you using?**

For example: Chrome

**Reproduction URL**

I was not able to create a minimal reproduction, yet. Will update once I have one. Wanted to share the issue now so others can search for it and help.

**Describe your issue**

I am testing React 19, Next 15, React Compiler in our Project. When rendering our app I am getting multiple console errors in the browser like…

```
Warning: Invalid prop `data-headlessui-state` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.
    at c (webpack-internal:///(app-pages-browser)/./node_modules/@headlessui/react/dist/internal/open-closed.js:17:18)
    at u (webpack-internal:///(app-pages-browser)/./node_modules/@headlessui/react/dist/internal/close-provider.js:13:18)
    at me (webpack-internal:///(app-pages-browser)/./node_modules/@headlessui/react/dist/components/disclosure/disclosure.js:111:24)
    at SelectDatasets (webpack-internal:///(app-pages-browser)/./src/app/regionen/[regionSlug]/_components/SidebarLayerControls/StaticDatasets/SelectDatasets.tsx:32:72)
    at nav
    …
```

I am seeing the same error with other libraries, so this might be a temporary or a new issue with React 19 or Next 15 or React Compiler.

**Comment by RobinMalfait** [comment]
Hmm interesting, I'm assuming you are using Headless UI in `SelectDatasets`. Do you have access to the generated output that the compiler outputs for this file?

**Comment by tordans** [comment]
> I'm assuming you are using Headless UI in `SelectDatasets`. 

Yes, this case is from https://github.com/FixMyBerlin/atlas-app/blob/react-compiler-react-19/src/app/regionen/%5BregionSlug%5D/_components/SidebarLayerControls/StaticDatasets/SelectDatasets.tsx#L1-L8 and starts in the `Disclosure`. However there are quite a few similar warnings from other components.

> Do you have access to the generated output that the compiler outputs for this file?

I am seeing this value `2` on which  I assume is the Fragment `u`: 
<img width="1181" alt="image" src="https://github.com/tailwindlabs/headlessui/assets/111561/2800608b-8874-4eea-b23b-5758a8783891">

Any recommendation how to extract the relevant code?

**Comment by kuskhan** [comment]
I have the same problem. It only occurs only in development mode.

**Comment by agusabas** [comment]
same issue testing React 19 rc.. waiting for updates!

**Comment by 6ichem** [comment]
Same issue happening in React 19 RC

Solution that worked for me is to add the `as="div"` prop to the `<Disclosure />` component like this: `<Disclosure {...rest} as="div">`

**Comment by yuriiIykyknow** [comment]
> Same issue happening in React 19 RC
> 
> Solution that worked for me is to add the `as="div"` prop to the `<Disclosure />` component like this: `<Disclosure {...rest} as="div">`

don't work for me

**Comment by michaelkhalsa** [comment]
Thanks @6ichem   
this worked for me as well by adding the as="div" prop to the Disclosure wrapper component.  
Wanted to add, that on React 18.2

**Comment by xsjcTony** [comment]
Happening in React 19 with Vite 6, happening in `Listbox`, causing the whole page to be frozen.

**Comment by mblazek-moravio** [comment]
Happening with React 19 and Next 15.1.6 when using Menu. Using `as="div"` resolved the issue. Any updates if this gets fixed?

**Comment by RobinMalfait** [comment]
Hey!

Can anyone running into this issue provide a minimal reproduction repo? Inside of Headless UI we check whether a `Fragment` was used or not. Only thing I can think of is that there is a different instance of `Fragment`, but looking at the React Compiler it seems like they just use `Fragment` from the `react` package directly.

Would love to get it fixed, so if anyone can share a reproduction that will be greatly appreciated.

**Comment by alekseikurnosenko** [comment]
This can be mitigated by replacing the fragment with `<div>` with `display: contents` style

**Comment by psam44** [comment]
I'm not a user of this repo, but I have a similar problem with react-native:
```
Warning: Invalid prop onLayout supplied to React.Fragment.
```
I filed the issue 50817 at https://github.com/facebook/react-native.

What looks similar to me is in:
headlessui/packages/@headlessui-react/src/utils/render.ts

```
    if (exposeState) {
      dataAttributes['data-headlessui-state'] = states.join(' ')    <-- HERE
...
        return cloneElement(
          resolvedChildren,    <-- HERE, is probably the Fragment
          Object.assign(
            {},
            mergedProps,
            dataAttributes,    <-- HERE, prop addition
...
```

I don't see the relation with the packages/react-native/Libraries/Renderer/implementations/ReactFabric-dev.js of my case (is Fabric an engine library used in both products?).

**Comment by david-crespo** [comment]
@psam44 your issue is related only in that the library you’re using is violating the same prohibition as this one. The prohibition comes from React itself. You can reproduce this with plain React without any libraries by making a fragment and passing props other than `key` and `children` to it.

**Comment by psam44** [comment]
If you look closely at my comment in 50817, I know who is the producer of the 'onLayout' prop.
But I located the exact error message in the Fabric renderer (so before React, I suppose).
May be there is the same checker in React, but I didn't spot something like that in the source repo (and I don't really know where to search for).
If you know where the check is in React, I'm interested.

**Comment by psam44** [comment]
Finally found in https://github.com/facebook/react/blob/main/packages/react-reconciler/src/ReactChildFiber.js

```
function validateFragmentProps(
...
      if (
        key !== 'children' &&
        key !== 'key' &&
        (enableFragmentRefs ? key !== 'ref' : true)
      ) {
...
            if (enableFragmentRefs) {
              console.error(
                'Invalid prop `%s` supplied to `React.Fragment`. ' +
                  'React.Fragment can only have `key`, `ref`, and `children` props.',
                erroredKey,
              );
            } else {
              console.error(
                'Invalid prop `%s` supplied to `React.Fragment`. ' +
                  'React.Fragment can only have `key` and `children` props.',
                erroredKey,
              );
            }
...
```

**Comment by psam44** [comment]
My issue for react-native has now a fix committed. And the decision is to inhibit the producer to provide additional props when the child is a Fragment.
Extract:
```
    const isFragment = element.type === React.Fragment;
    if (isFragment) {
      return element;
    }
    return React.cloneElement(element, {
      onLayout: (event: LayoutChangeEvent) => {
...
```

So for your ListBox, I suppose you have to choose: either accept to not support a Fragment as a child, either do not pass any additional attribute (i.e. return `resolvedChildren` as is).

**Comment by ikramhasib007** [comment]
> Same issue happening in React 19 RC
> 
> Solution that worked for me is to add the `as="div"` prop to the `<Disclosure />` component like this: `<Disclosure {...rest} as="div">`

Thank you. It solved!
